AnderCelsius/MarthasLibrary
# Add Comprehensive Integration Tests for Books Controller Actions

Title: Add Comprehensive Integration Tests for Books Controller Actions

---

## Summary
This Pull Request introduces a suite of integration tests covering all the actions in the `BooksController`. The purpose of these tests is to ensure that the controller behaves as expected with various inputs and correctly interacts with the underlying services and the database. These tests validate the expected outcomes of the endpoints, including success, failure, and edge cases.

## Changes
- Added integration tests for the `Create` action to verify book creation with valid and invalid data.
- Implemented tests for the `GetAll` and `GetById` actions to check the retrieval of books/book under different scenarios.
- Crafted tests for the `Search` action to ensure that the search functionality returns the correct books based on a search term and handles empty queries.
- Included tests for the `Update` and `Delete` actions to confirm they correctly modify or remove book entries.
- Set up a shared `TestFixture` to initialize the testing environment, which includes setting up an in-memory database and mocking external services where necessary.
- Configured GitHub Actions workflow to run integration tests automatically on push and pull requests.

## Test Strategy
The tests are designed to cover:
- **Happy path scenarios**: Confirming that the API responds correctly with the expected success codes and data formats when provided with valid input.
- **Validation failures**: Ensuring that the API responds with appropriate error messages and status codes when provided with invalid input.
- **Edge cases**: Testing the API's response to uncommon or extreme inputs, including boundary values and empty/null inputs.

## Does not include
- **Authorization checks**: This PR does not verify that the API properly restricts access to endpoints based on authentication and authorization.

## How to Test
To run the integration tests, follow these steps:
1. Ensure that the latest code is pulled from the repository.
2. Navigate to the integration tests directory.
3. Run the command `dotnet test` to execute the integration tests.

The tests will automatically use the `TestFixture` to set up a test environment, so no additional configuration should be necessary.

## Dependencies
No additional dependencies have been introduced. The tests utilize the existing project references and rely on the `Microsoft.EntityFrameworkCore.InMemory` provider for database operations.

## Screenshots
_N/A - This PR does not include visual changes._

## Additional Notes
- The integration tests assume that the `BooksController` is already fully implemented.
- Any existing unit tests have been left untouched and continue to operate as expected.
- Future work could include adding authorization checks.
